artiq/coredevice/comm_kernel.py

```diff
@@ -264,7 +271,11 @@ def _read_string(self):
     #
 
     def _write(self, data):
-        self.socket.sendall(data)
+        # make the send non-blocking, so large buffer send and computation can
+        # be done in parallel
+        # we have to wait before sending the next time
+        select.select([], [self.socket], [])
+        self.socket.sendall(data, socket.SOCK_NONBLOCK)
```

Comment:
There is probably some problem there.

If ``socket.sendall`` with ``SOCK_NONBLOCK`` can accept arbitrarily large buffers, why would it ever block in ``select``?

I suspect that either:
1. ``select`` never blocks and is not necessary, ``sendall`` has an arbitrarily large memory buffer.
2. ``sendall``  when used with ``SOCK_NONBLOCK`` actually blocks when the buffer is full, and ``select`` is not necessary,
3.  some data is dropped by ``sendall`` when used with ``SOCK_NONBLOCK`` if the input data does not fit in the available buffer space.

Reply:
Also I wonder if the behavior is consistent between operating systems and Windows.

Reply:
I am not very sure about the behavior of `sendall` for a non-blocking socket, the documentation is not very clear. I think I would try and see if the `select` is really needed here. And yes, Windows is a concern to me, I think if we use `select`, we should test it on Windows before merging.

Reply:
Yep, maybe reading the source is necessary here.